src/Data/HashMap/Linear.hs

```diff
@@ -59,218 +61,219 @@ import qualified Prelude
   -- | A key, PSL pair to insert at an index, that has not
   -- yet been touched. Invariant: the PSL must be -1 at that index.
-  IndexToInsert :: (k, PSL) -> Int -> RobinQuery k
+  IndexToInsert :: (k, PSL) -> Int -> ProbeResult k
   -- A key, PSL pair to update the value at an index.
-  IndexToUpdate :: (k, PSL) -> Int -> RobinQuery k
+  IndexToUpdate :: (k, PSL) -> Int -> ProbeResult k
   -- | A key, PSL pair for a swap at an index.
   -- The swapped-out pair will then need to be inserted downstream.
-  IndexToSwap :: (k, PSL) -> Int -> RobinQuery k
+  IndexToSwap :: (k, PSL) -> Int -> ProbeResult k
+
 
 -- # Construction and Modification
 --------------------------------------------------
 
--- | Run a computation using a singleton hashmap
-singleton :: Keyed k =>
-  (k, v) -> (HashMap k v #-> Unrestricted b) -> Unrestricted b
-singleton (k :: k, v :: v) (f :: HashMap k v #-> Unrestricted b) =
-  alloc defaultSize (k, v, -1) applyHM
-  where
-    applyHM :: Array (RobinVal k v) #-> Unrestricted b
-    applyHM arr = let ixToHash = (hash k) `mod` defaultSize in
-      f $ HashMap (defaultSize, 1) (write arr ixToHash (k, v, 0))
-
-singleton' :: Keyed k =>
-  (k, v) -> (HashMap k v #-> Unrestricted b) -> b
-singleton' kv f = unUnrestricted (singleton kv f)
-
--- XXX: re-write linearly
--- | Given a key @k@ to 'lookup', look up a @Maybe v@ and feed it to
--- a function @Maybe v -> Maybe v@, and if the result is @Nothing@,
--- delete that key, and otherwise, replace the value with the @v@
--- in the returned @Just v@.
-alter ::
-  Keyed k =>
-  HashMap k v #-> (Maybe v -> Maybe v) ->
-  k ->
-  HashMap k v
-alter = Unsafe.toLinear unsafeAlter
+-- | Run a computation given an empty hashmap
+empty :: forall k v b.
+  Keyed k => Size -> (HashMap k v #-> Unrestricted b) -> Unrestricted b
+empty sz@(Size s) scope =
+  alloc
+    s
+    ( -1
+    , ( error "reading error hashmap key"
+      , error "reading error hashmap val"
+      )
+    )
+    scope'
   where
-    unsafeAlter :: Keyed k =>
-      HashMap k v -> (Maybe v -> Maybe v) -> k -> HashMap k v
-    unsafeAlter hmap f k =
-      case lookup hmap k of
-        (hmap', Unrestricted maybeV) -> case f maybeV of
-          Nothing -> delete hmap' k
-          Just v -> insert hmap' k v
-
--- | If the key is present, this update the value,
--- otherwise insert a new key-value pair.
-insert :: Keyed k => HashMap k v #-> k -> v -> HashMap k v
-insert hmap k v = insertFromQuery v $ queryIndex (maybeResize hmap) k
+    scope' :: RobinArr k v #-> Unrestricted b
+    scope' arr = scope $ HashMap sz (Count 0) arr
+
+-- | If the key is present, this update the value.
+-- If not, if there's enough space, insert a new pair.
+-- If there isn't enough space, resize and insert
+insert :: (Keyed k, HasCallStack) => HashMap k v #-> k -> v -> HashMap k v
+insert (HashMap sz@(Size s) ct@(Count c) arr) k v = case c < s of
+  False -> resizeMap (HashMap sz ct arr) & \case
+    !hmap' -> insert hmap' k v
```

Comment:
It's not entirely clear from the context why you are using a `!` pattern here. But it looks like it may be very important. So add a comment explaining why you are forcing your result here. For the benefit of the future generations.

Reply:
Whoops, left that in there -- there's no reason, just made it easier to put in print statements for debugging.